mlir_graphblas/src/lib/GraphBLAS/GraphBLASLowerPass.cpp

```diff
@@ -1669,6 +1683,191 @@ class LowerMatrixMultiplyRewrite
   };
 };
 
+class MatrixMultiplyGenericDWIMFirstArgRewrite
+    : public OpRewritePattern<graphblas::MatrixMultiplyGenericOp> {
+public:
+  using OpRewritePattern<graphblas::MatrixMultiplyGenericOp>::OpRewritePattern;
+
+  template <class T>
+  static bool needsDWIM(T op) {
+    Value A = op.a();
+    RankedTensorType aType = A.getType().cast<RankedTensorType>();
+    int64_t aRank = aType.getRank();
+    if (aRank == 2) {
+      bool aIsCSC = hasColumnOrdering(aType);
+      if (aIsCSC) {
+        return true;
+      }
+    }
+    return false;
+  };
+
+  LogicalResult match(graphblas::MatrixMultiplyGenericOp op) const override {
+    if (needsDWIM(op))
+      return success();
+    else
+      return failure();
+  };
+
+  void rewrite(graphblas::MatrixMultiplyGenericOp op,
+               PatternRewriter &rewriter) const override {
+    MLIRContext *context = op.getContext();
+    Location loc = op->getLoc();
+
+    Value A = op.a();
+    Value B = op.b();
+    Value mask = op.mask();
+    bool maskComplement = op.mask_complement();
+
+    RankedTensorType aType = A.getType().cast<RankedTensorType>();
+    RankedTensorType flippedMatrixType = getFlippedLayoutType(context, aType);
+    Value flippedA =
+        rewriter.create<graphblas::ConvertLayoutOp>(loc, flippedMatrixType, A);
+
+    NamedAttrList attributes = {};
+    attributes.append(StringRef("mask_complement"),
+                      rewriter.getBoolAttr(maskComplement));
+
+    graphblas::MatrixMultiplyGenericOp newOp;
+    if (mask)
+      newOp = rewriter.create<graphblas::MatrixMultiplyGenericOp>(
+          loc, op->getResultTypes(), ValueRange{flippedA, B, mask},
+          attributes.getAttrs(), 3);
+    else
+      newOp = rewriter.create<graphblas::MatrixMultiplyGenericOp>(
+          loc, op->getResultTypes(), ValueRange{flippedA, B},
+          attributes.getAttrs(), 3);
+
+    newOp.getRegion(0).takeBody(op.getRegion(0));
+    newOp.getRegion(1).takeBody(op.getRegion(1));
+    newOp.getRegion(2).takeBody(op.getRegion(2));
+
+    rewriter.replaceOp(op, newOp.getResult());
```

Comment:
This feels very verbose to create a new op with mostly the same arguments. If there isn't a way to write this more compactly, we should create a utility function to avoid so much duplicated code.

Reply:
I found a nice way to replace an operand rather than a whole operation.